### PR TITLE
Dropimportnotstrict492 - More updates to get things through rocket-chip.

### PR DESCRIPTION
**Pull request by ucbjrl**
Add missing implicit compileOptions to do_pad() and do_zext().
Add SourceInfoTransform macros to Vec methods in order to avoid apply chain issues.
Since utils methods are no longer NotStrict, Pipe objects need access to the client's compile options. There may be more.

### CODE REVIEW

File: src/main/scala/chisel3/util/Valid.scala [hunk 1]
```diff
@@ -34,7 +36,7 @@ object Valid {
 object Pipe
 {
   @chiselName
-  def apply[T <: Data](enqValid: Bool, enqBits: T, latency: Int): Valid[T] = {
+  def apply[T <: Data](enqValid: Bool, enqBits: T, latency: Int)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): Valid[T] = {
```
Comment (ducky64) [inline src/main/scala/chisel3/util/Valid.scala]: Source info shouldn't be propagated through helper functions, only chisel internals.
Source info support for libraries is something that needs more thought and discussion.

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala [hunk 2]
```diff
@@ -243,11 +246,11 @@ sealed class Vec[T <: Data] private (gen: => T, val length: Int)
   def apply(idx: Int): T = self(idx)
 
   @deprecated("Use Vec.apply instead", "chisel3")
-  def read(idx: UInt)(implicit compileOptions: CompileOptions): T = apply(idx)
+  def read(idx: UInt)(implicit compileOptions: CompileOptions): T = do_apply(idx)(DeprecatedSourceInfo, compileOptions)
 
   @deprecated("Use Vec.apply instead", "chisel3")
   def write(idx: UInt, data: T)(implicit compileOptions: CompileOptions): Unit = {
-    apply(idx)(compileOptions).:=(data)(DeprecatedSourceInfo, compileOptions)
+    do_apply(idx)(DeprecatedSourceInfo, compileOptions).:=(data)(DeprecatedSourceInfo, compileOptions)
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Can we just not thread compile options through deprecated functions?
These should be removed eventually anyways.

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala [hunk 3]
```diff
@@ -223,7 +223,10 @@ sealed class Vec[T <: Data] private (gen: => T, val length: Int)
 
   /** Creates a dynamically indexed read or write accessor into the array.
     */
-  def apply(idx: UInt)(implicit compileOptions: CompileOptions): T = {
+  override def apply(x: UInt): T = macro SourceInfoTransform.xArg
+
+  def do_apply(x: UInt)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions): T = {
+    val idx = x
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: You could also just change all instances of `idx` to `x`.

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala (hunk 2 above)
Comment (ucbjrl) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: Thanks @ducky64. I hope d9eff59 addresses this.

**Pull request merged**